vault/eventbus/bus.go

```diff
@@ -0,0 +1,158 @@
+func (bus *EventBus) Start() {
+	bus.started.Store(true)
+}
+
+var _ logical.EventSender = (*EventBus)(nil)
+
+// Send sends an event to the event bus and routes it to all relevant subscribers.
+// This function does *not* wait for all subscribers to acknowledge before returning.
+// TODO: use schema once it is defined
+func (bus *EventBus) Send(ctx context.Context, eventType string, s any) error {
+	if !bus.started.Load() {
+		return ErrNotStarted
+	}
+	bus.logger.Info("Sending event", "event", s)
+	_, err := bus.broker.Send(ctx, eventlogger.EventType(eventType), s)
+	if err != nil {
+		// if no listeners for this event type are registered, that's okay, the event
+		// will just not be sent anywhere
+		if strings.Contains(strings.ToLower(err.Error()), "no graph for eventtype") {
+			return nil
+		}
+	}
+	return err
+}
+
+func (bus *EventBus) Subscribe(_ context.Context, eventType string) (chan any, error) {
+	// subscriptions are still stored even if the bus has not been started
+	pipelineID, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	nodeID, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+	formatterNodeID, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: should we have just one node, and handle all the routing ourselves?
+	asyncNode := newAsyncNode()
+	err = bus.broker.RegisterNode(eventlogger.NodeID(nodeID), asyncNode)
+	if err != nil {
+		defer asyncNode.Close()
+		return nil, err
+	}
+
+	err = bus.broker.RegisterNode(eventlogger.NodeID(formatterNodeID), cloudEventsFormatterFilter)
+	if err != nil {
+		defer asyncNode.Close()
+		return nil, err
+	}
+
+	nodes := []eventlogger.NodeID{eventlogger.NodeID(formatterNodeID), eventlogger.NodeID(nodeID)}
+
+	pipeline := eventlogger.Pipeline{
+		PipelineID: eventlogger.PipelineID(pipelineID),
+		EventType:  eventlogger.EventType(eventType),
+		NodeIDs:    nodes,
+	}
+	err = bus.broker.RegisterPipeline(pipeline)
+	if err != nil {
+		defer asyncNode.Close()
+		return nil, err
+	}
+	return asyncNode.ch, nil
+}
+
+var cloudEventsFormatterFilter *cloudevents.FormatterFilter
+
+func init() {
+	// TODO: maybe this should relate to the Vault core somehow?
+	sourceUrl, err := url.Parse("https://vaultproject.io/")
+	if err != nil {
+		panic(err)
+	}
+	cloudEventsFormatterFilter = &cloudevents.FormatterFilter{
+		Source: sourceUrl,
+		Predicate: func(_ context.Context, e interface{}) (bool, error) {
+			return true, nil
+		},
+	}
+}
+
+type asyncChanNode struct {
+	// TODO: add bounded deque buffer of *any
+	ch chan any
+}
+
+func newAsyncNode() *asyncChanNode {
+	return &asyncChanNode{
+		ch: make(chan any),
+	}
+}
+
+func (node *asyncChanNode) Close() error {
+	close(node.ch)
+	return nil
+}
+
+func (node *asyncChanNode) Process(ctx context.Context, e *eventlogger.Event) (*eventlogger.Event, error) {
+	// TODO: add timeout on sending to node.ch
+	go func() {
+		select {
+		case node.ch <- e.Payload: // sends to the channel async
```

Comment:
This comment confused me because the channel is unbuffered, so this is actually a synchronous send. The thing that makes it async is being inside a goroutine.

Reply:
Ah, I can move and clarify the comment.